File: c/binary-prefix-divisible-by-5.c
/**
 * 1018. 可被 5 整除的二进制前缀：https://leetcode.cn/problems/binary-prefix-divisible-by-5
 */

#include <stdbool.h>
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
bool* prefixesDivBy5(int* nums, int numsSize, int* returnSize) {
    *returnSize = numsSize;
    bool* prefixes = malloc(sizeof(bool) * numsSize);

    int x = 0;
    for (int i = 0; i < numsSize; i++) {
        x = ((x << 1) | nums[i]) % 5;
        prefixes[i] = x == 0;
    }
    return prefixes;
}

File: java/src/main/BinaryPrefixDivisibleBy5.java
import java.util.ArrayList;
import java.util.List;

/**
 * 1018. 可被 5 整除的二进制前缀：https://leetcode.cn/problems/binary-prefix-divisible-by-5
 */
public class BinaryPrefixDivisibleBy5 {

    public List<Boolean> prefixesDivBy5(int[] nums) {
        List<Boolean> res = new ArrayList<>(nums.length);
        int x = 0;
        for (int num : nums) {
            x = ((x << 1) + num) % 5;
            res.add(x == 0);
        }
        return res;
    }

}
